### ISSUE TITLE
Attention Mechanism in Encoder and Decoder

### ISSUE THREAD

**Issue by timechild666** [Zenome84/tft_keras#1]
Thank for sharing your Keras implementation of the TFT. I really love the idea of having a more compact and up to date Keras version of the model. As I was examining the code, I noticed that the model doesn't seem to include attention mechanisms in both the encoder and decoder parts. As far as I understand, the original TFT model should have self-attention in both the encoder and decoder components.

I wanted to point this out and seek clarification on whether this was an intentional design decision or if the implementation is still a work in progress. Including attention in both components might improve the model's performance, especially for complex time series forecasting tasks.

I might be wrong on this, so I'd appreciate it if you could shed some light on the subject.

I appreciate your work and I'm looking forward to your response.

Martin

**Comment by Zenome84** [comment]
@timechild666 thanks for your interest and questions. I will try to answer each one:

> I wanted to point this out and seek clarification on whether this was an intentional design decision or if the implementation is still a work in progress.

So it is still a work in progress (WIP) with work left present in the TODO found in the [README.md file](https://github.com/Zenome84/tft_keras/blob/main/README.md). However, as I will explain further, the implementation is true to the original work.

> As I was examining the code, I noticed that the model doesn't seem to include attention mechanisms in both the encoder and decoder parts.

Both encoder and decoder networks feed into the self-attention mechanism. Notice that I concatenate both [here](https://github.com/Zenome84/tft_keras/blob/642f9cc06f213e0dac6a2bb12b79602d282941d4/tft/tft.py#L144): `lstm_layer = tf.keras.layers.concatenate([observed_lstm, forecast_lstm], axis=1)`. Notice I use the named convention of observed for the encoder and forecast for the decoder - I find that this is more descriptive than encoder/decoder.
This combined layer then becomes the `enriched_temporal_layer` that goes into the `InterpretableMultiHeadSelfAttention`.

I hope this clarifies things. While the module is a WIP, the actual network is fully completed to the best of my ability to review my own code.
At the moment, you can call the internal model to use it, and write a custom quantile loss function:
```
tft = TemporalFusionTransformer(...)
tft.model.fit(...)
```
I have tested it on a dataset for a different project that I am able to progress every month or two when I find 3-8 hours of time to dedicate to it. Hopefully, my next burst of work will come soon :)

**Comment by timechild666** [comment]
I would like to express my gratitude for the detailed explanation and clarification on the subject.

I understand that the implementation is a work in progress, but I do have an additional question about the attention mechanism used in the TFT Keras implementation. The original TFT code utilizes Scaled Dot Product Attention in conjunction with the Interpretable Multi-Head Attention mechanism. In the current TFT Keras implementation, I see that you are employing the Keras Attention layer within the InterpretableMultiHeadSelfAttention layer.

I noticed in the code you shared that the query is divided by the square root of the key's dimension size in the call() method, which seems to incorporate the Scaled Dot Product Attention. Correct me if I'm wrong here, but can you please provide clarification on whether the Scaled Dot Product Attention is already implemented in the current code?

Thank you for your time!

**Comment by Zenome84** [comment]
Sure, let me forward you to the [TF documentation on Attention](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Attention). Notice that the default implementation uses the dot product. This is true to formula (9) in the original paper.

Unfortunately, the Keras layer does not provide the fixed scaling option in the paper, so I simply perform this scaling on the input.

> I would like to express my gratitude for the detailed explanation and clarification on the subject.

I welcome the questions as it provides a chance for review.
